### ISSUE TITLE
Do not include <winsock2.h> from <exiv2/exiv2.hpp>

### ISSUE THREAD

**Issue by clanmills** [Exiv2/exiv2#1335]
https://bitbucket.org/agriggio/art/issues/128/msys2-compiling-errors-with-windows-10-and#comment-58935959

**Comment by tester0077** [comment]
Hi Robbin,

FWIW, though I haven't followed this thread nor the one relating to the 
rest of the winsock.h issues, I know I had to recently add a line to the 
very top of my code for several files which (eventually) include the 
exiv2.hpp header

#define WIN32_LEAN_AND_MEAN   // needed for new Exiv2 lib

that line needs to be ahead of any other #include line which includes 
any Windows header(s)

Don't know if this might help or fix the issue or if that has been 
tested/tried before

Arnold

**Comment by clanmills** [comment]
Thanks, Arnold.  Yes WIN32_LEAN_AND_MEAN.  That's some kind of throttle to reduce the stuff being pulled in.  Perhaps I should be declaring that before I pull in <windows.h>.  I don't know.  When I make a release, I know that the exiv2 library and sample applications build and pass the test suite.  By publishing the Release Candidates, I hope that users will report interactions between exiv2 and their applications.  Occasionally, they do.  Mostly they don't look at the release until it has been released.

In this case, I promoted <winsock2.h> for my convenience and didn't think about it have an impact on other folks code.

If you need WIN32_LEAN_AND_MEAN, could you open an issue about it please and I'll deal with it.  I'm surprised to see <windows.h> is included in quite a number of files.  I'm sure that can also be simplified:

```bash
991 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ finder "*pp" | xargs grep windows.h 
./include/exiv2/rwlock.hpp:#include <windows.h>
./samples/exiv2json.cpp:#include <windows.h>
./samples/geotag.cpp:#include <windows.h>
./xmpsdk/src/XMPCore_Impl.hpp:	#include <windows.h>
./src/types.cpp:# include <windows.h> // for MultiByteToWideChar etc
./src/jpgimage.cpp:#include <windows.h>
./src/convert.cpp:# include <windows.h>
./src/version.cpp:# include <windows.h>
./src/basicio.cpp:# include <windows.h>
./src/exiv2app.hpp:#include <windows.h>
./src/makernote_int.cpp:#include <windows.h>
./src/futils.cpp:#include <windows.h>
992 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ 
```

The Exiv2 project is killing me.  I want to get on with the book.  Many days bring 8 hours of unscheduled work.

**Comment by tester0077** [comment]
Hi Robin

On 2020-10-01 2:18 PM, Robin Mills wrote:
>
> Thanks, Arnold. Yes WIN32_LEAN_AND_MEAN. That's some kind of throttle 
> to reduce the stuff being pulled in. Perhaps I should be declaring 
> that before I pull in <windows.h>. I don't know.
>
I have had to add that 'rider' in a number of my programs including some 
which use Exiv2lib

And, yes, IMO, you need to add the #define before including windows.h

> When I make a release, I know that the exiv2 library and sample 
> applications build and pass the test suite.
>
Well, again, IMO, unless the changes fixes a definite issue, the test 
suite is an internal test and does not really address the issues your 
user's might & likely will face :-(

I would expect that most of the test cases you write are fine-tuned 
until everything passes and accepts the latest changes to the code. 
Useful for you, but not so much for the users - they always use thing 
differently and usually make quite different assumptions ;-)

The real proof of the pudding is in the user's consuming and using the code

> By publishing the Release Candidates, I hope that users will report 
> interactions between exiv2 and their applications. Occasionally, they 
> do. Mostly they don't look at the release until it has been released.
>
Just speaking for myself, (and the comments in the thread by the 
originator seems to bear me out) I don't and simply do not have the time 
to test all release candidates.

In fact, from my experience, I need to build my own local copy, mainly 
because some functions within Exiv2lib conflict with my long established 
and trustworthy memory leak detection code. FWIW, Visual Leak Detector 
does not seem to be usable with the latest M$ compilers etc. last time I 
checked, while the M$ code I am using has been reliably catching leaks 
for me for as far back as MSVC 2010, IIRC.

Other than that, some of the most recent changes (AutoPtr/UniquePtr 
comes to mind) make me very reluctant to update simply because a 
somewhat newer version or release candidate has been announced/released. 
Updating will most likely cost me time, without necessarily giving me 
something I cannot do without.

> In this case, I promoted <winsock2.h> for my convenience and didn't 
> think about it have an impact on other folks code.
>
Actually, I am quite used to this issue and so I typically have no 
problem dealing with it. Still, when I cloned the latest Exiv2 code from 
Github, the issue popped up, but caused me little lost time
>
> If you need WIN32_LEAN_AND_MEAN, could you open an issue about it 
> please and I'll deal with it.
>
Why not simply use the already open issue #1335 as the trigger?
>
> I'm surprised to see <windows.h> is included in quite a number of 
> files. I'm sure that can also be simplified:
>
Well, of course, all this is only a problem under Windows and M$ has 
crammed way too much into the windows.h header and then us users have to 
fudge things to make our code work
>
> The Exiv2 project is killing me. I want to get on with the book. Many 
> days bring 8 hours of unscheduled work.
>
You have my sympathy, for all the good that will do :-(

Though, I am sure the book will be the better for this extra work.

Both the book and your work in making Exiv2 usable, go hand in hand, 
IMNSHO :-)

Arnold

**Comment by clanmills** [comment]
I haven't been able to do any work on the book this week.  I think we've almost reached the end of rewriting the bash scripts in python.  I'll try to clean up all this unexpected windows stuff today.  Monday, I will focus 100% on the book.

I'm not going to do anything about WIN32_LEAN_AND_MEAN until somebody reports a solid reason to do something about it.   It's a common thing for windows programs to use.  Something like:

```cpp
#define WIN32_LEAN_AND_MEAN
#include <windows.h>
....
#include <exiv2/exiv2.hpp>
....
```

**Comment by clanmills** [comment]
Oh, by the way, it's Robin, not Robbin.  My mother's name was Robina (a Victorian name) and I was named after her.  She was always called "Beanie".  Robin is a very common boy's first name in Scotland because I believe Robert Burns, the Scottish National Poet, was actually called Robin.  He wrote http://www.robertburns.org/works/65.shtml

There was a lad was born in Kyle, 
But whatna day o' whatna style, 
I doubt it's hardly worth the while 
To be sae nice wi' Robin. 

_Chorus:_
Robin was a rovin' boy, 
Rantin', rovin', rantin', rovin', 
Robin was a rovin' boy, 
Rantin', rovin', Robin!

**Comment by clanmills** [comment]
Fixed with #1340.  Pleased to say the CI is working fine and required no attention.

**Comment by piponazo** [comment]
At the moment we have the following CMake code in the file `cmake/compilerFlags.cmake`:

```
    add_definitions(-DNOMINMAX -DWIN32_LEAN_AND_MEAN)
```

That's adding the mentioned compiler definition for windows builds when building Exiv2, but it is not propagated to consumers. There is a way to do exactly that with CMake, and I want to take care of this (that's why I will reopen this issue). I also want to take care of the following points discussed:

* Reduce the numbers of inclusions of the "windows.h" header by moving it to the config.h header.
* Remove the inclusion of "winsock2.h" from the config.h header and only use it where it is needed (http.cpp).
* Remove the `EXIV2_BUILDING_EXIV2` mechanism.

**Comment by clanmills** [comment]
Thanks for picking up this one.  I think you have a username/password for the MacMini/buildserver if you need windows/cygwin/mingw build machines.  I'm happy to help you access and use that if you wish.

The CI doesn't build Cygwin and MinGW/msys2.  I'm not trying to dump work on you, however it would be really good to support that.  I have been unable to get conan to work on those platforms.  There are instructions in README.m about setting up the dependencies on Cygwin and MinGW/msys2.

**Comment by clanmills** [comment]
Fixed and closed in #1349.
